### DISCUSSION TITLE
loosing the type on some funtions

### DISCUSSION THREAD

**Discussion by joacoiannuzzi** [orval-labs/orval#165] [General]
Hello, first of all I want to thank you for this awesome library.

I am having a problem that in some generated functions I am loosing the type of the response.

```ts
export const loginUsingPOST = <Data = unknown>(
  logInUserDTO: LogInUserDTO,
  options?: SecondParameter<typeof customInstance>
) => {
  return customInstance<Data extends unknown ? UserDTO | unknown : Data>(
    { url: `/auth/login`, method: "post", data: logInUserDTO },
    // eslint-disable-next-line
    // @ts-ignore
    options
  );
};
```

As you can see the response type is <Data extends unknown ? UserDTO | unknown : Data>, which because of typescript, the whole thing becomes unknown.

This same thing I saw is happening in the mutation in the react query example.

Is there a way to fix this?

Thanks

**Comment by anymaniax** [comment]
Hello @joacoiannuzzi, can you share your swagger, please?

**Reply by joacoiannuzzi** [reply]
```json
"/auth/login": {
      "post": {
        "tags": [
          "auth-controller"
        ],
        "summary": "login",
        "operationId": "loginUsingPOST",
        "consumes": [
          "application/json"
        ],
        "produces": [
          "*/*"
        ],
        "parameters": [
          {
            "in": "body",
            "name": "logInUser",
            "description": "logInUser",
            "required": true,
            "schema": {
              "$ref": "#/definitions/LogInUserDTO"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "$ref": "#/definitions/UserDTO"
            }
          },
          "201": {
            "description": "Created"
          },
          "401": {
            "description": "Unauthorized"
          },
          "403": {
            "description": "Forbidden"
          },
          "404": {
            "description": "Not Found"
          }
        }
      }
    },
```

**Reply by anymaniax** [reply]
If I understand well, you create and log in your user with the same endpoint?

**Reply by anymaniax** [reply]
can you also try to regenerate with the last version. I updated the type gesture maybe it will help but I don't think in your case

**Reply by joacoiannuzzi** [reply]
> If I understand well, you create and log in your user with the same endpoint?

Actually no, the schema is auto generated from the backend api, and all mutation responses get that signature of having 200, 201, ...etc.


> can you also try to regenerate with the last version. I updated the type gesture maybe it will help but I don't think in your case
I tried it but no luck, the problem is still happening

```ts
export const loginUsingPOST = <TData = UserDTO | unknown>(
  logInUserDTO: LogInUserDTO,
  options?: SecondParameter<typeof customInstance>
) => {
  return customInstance<TData>(
    { url: `/auth/login`, method: "post", data: logInUserDTO },
    // eslint-disable-next-line
    // @ts-ignore
    options
  );
};
```

Also, I found another problem on the generated mocks. Most responses get made with the possibility of getting undefined. I think the problem is related to the other.

```ts
export const getLoginUsingPOSTMock = () =>
  faker.helpers.randomize([
    {
      email: (() => faker.internet.email())(),
      id: (() => faker.datatype.uuid())(),
      userName: faker.random.word(),
    },
    undefined,
  ]);
``` 

Thanks

**Reply by anymaniax** [reply]
Maybe be you can do a transformer like [here](https://github.com/anymaniax/orval/blob/master/samples/react-app-with-react-query/src/api/transformer/add-version.js) to automatically remove the 201 if you already have a 200. That the only way at the moment to remove a response type.

**Reply by joacoiannuzzi** [reply]
Thanks, I was able to do what you said, and removed the responses that didn't have any content.
Now I ended up with this:

```ts
export const loginUsingPOST = <TData = UserDTO>(
  logInUserDTO: LogInUserDTO,
  options?: SecondParameter<typeof customInstance>,
) => {
  return customInstance<TData>(
    { url: `/auth/login`, method: 'post', data: logInUserDTO },
    // eslint-disable-next-line
    // @ts-ignore
    options,
  );
};
```

However there is still a problem with the response in the mutation hook: 

```ts
export const useLoginUsingPOST = <
  TData = UserDTO,
  TError = unknown,
  TContext = unknown
>(options?: {
  mutation?: UseMutationOptions<
    AsyncReturnType<typeof loginUsingPOST>,
    TError,
    { data: LogInUserDTO },
    TContext
  >;
  request?: SecondParameter<typeof customInstance>;
}) => {
  const { mutation: mutationOptions, request: requestOptions } = options || {};

  return useMutation<
    AsyncReturnType<typeof loginUsingPOST>,
    TError,
    { data: LogInUserDTO },
    TContext
  >((props) => {
    const { data } = props || {};

    return loginUsingPOST<TData>(data, requestOptions);
  }, mutationOptions);
};
```
I think the problem is in the  "AsyncReturnType\<typeof loginUsingPOST\>", If you replace that with "TData", it works fine.

Also, I found out that in the msw mock responses functions, the errors responses of the schema are being used to generate the reponse data. That's why I was getting in all of them:  faker.helpers.randomize([ ..., undefined ]). 
I was able to overcome this by just removing all success or error responses in the schema that didn't have content.
Maybe it would be better to ignore the error responses in the mocks.

Thanks for the help!

**Reply by anymaniax** [reply]
For the problem with the mutate. I will fix it tonight normally that's an open issue

**Reply by anymaniax** [reply]
I published the `5.4.8` which should solve the problem with mutate
